ichnaea/api/views.py

```diff
@@ -120,6 +121,23 @@ def preprocess_request(self):
             # Use str(), since repr() includes the full source bytes
             raise self.prepare_exception(ParseError({"decode": str(exc)}))
 
+        if self.ip_log_and_rate_limit:
+            # Count unique daily requests
+            siggen = sha512(content.encode())
```

Comment:
A few weeks ago, I wanted to normalize inputs, but gave up on it. For this first round, I'm just looking at matching the bytes of the request body.

My first idea was that if two inputs looked the same to the location algorithms, then they should be treated as the same request. It appeared to me that the ``Query.json()`` method could be used to standardize the request, normalize whitespace, and discard unknown parameters. This would represent the ideal case of a client that assembles a request and then compares it to a request / response cache. However, I wasn't happy with how much ``Query.json()`` throws away, and there wasn't an equivalent for the submit API.

I then switched to more narrowly solve [bug 1624687](https://bugzilla.mozilla.org/show_bug.cgi?id=1624687). Could a naive proxy that doesn't try to understand the payload return a cached result? It seemed like a good trade-off between simplicity of implementation and usefulness of results. That resulted in the code in this PR.

This PR give us one ratio of unique to repeated requests. Normalizing the JSON content may move some requests from the unique to the repeated count.

If you think it is important to factor out order and whitespace differences in the JSON content, I could move the signature code a few lines down, and feed the re-serialized JSON to the hashing function. That would get us closer to what a smart client would send.

If you want to factor out invalid content as well, we could re-serialize the output of validating against the schema.

Reply:
Ok, so to answer my question, seems like we can't rely on clients to send requests exactly the same way every time.

Bug 1624687 seems to cover two different things. First, it seeks to measure the impact of caching where we want to cache _any_ incoming request regardless of whether it's valid or invalid--all are equally cacheable. I think this is fine for that purpose.

Second, it seeks to determine the impact of location caching in the Firefox client. I don't think these changes help us figure out the impact of location caching. Why can't we flatten and sort the parameters into a stable list before hashing?

Reply:
I agree, some stabilization would answer the question about caching in Firefox more directly. Should we stabilize before or after schema validation?

Reply:
After we have this data, are we able to filter it by the Firefox api keys? If so, then maybe we can assume that the requests Firefox is sending are valid schema-wise and thus either answer is fine.

Reply:
Yes, you can filter by API keys. I'll move and update the signature code.